src/rgw/rgw_sal_rados.h

```diff
@@ -632,30 +652,130 @@ class RadosGCChain : public GCChain {
     virtual void delete_inline(const DoutPrefixProvider *dpp, const std::string& tag) override;
 };
 
-class RadosWriter : public Writer {
+class RadosAtomicWriter : public Writer {
+protected:
   rgw::sal::RadosStore* store;
-  RGWSI_RADOS::Obj stripe_obj; // current stripe object
-
- public:
-  RadosWriter(Aio* aio, rgw::sal::RadosStore* _store,
-	      rgw::sal::Bucket* bucket,
-              RGWObjectCtx& obj_ctx, std::unique_ptr<rgw::sal::Object> _head_obj,
-              const DoutPrefixProvider* dpp, optional_yield y)
-    : Writer(aio, bucket, obj_ctx, std::move(_head_obj), dpp, y), store(_store)
-  {}
+  std::unique_ptr<Aio> aio;
+  rgw::putobj::AtomicObjectProcessor processor;
 
-  ~RadosWriter();
+public:
+  RadosAtomicWriter(const DoutPrefixProvider *dpp,
+		    optional_yield y,
+		    std::unique_ptr<rgw::sal::Object> _head_obj,
+		    RadosStore* _store, std::unique_ptr<Aio> _aio,
+		    const rgw_user& owner, RGWObjectCtx& obj_ctx,
+		    const rgw_placement_rule *ptail_placement_rule,
+		    uint64_t olh_epoch,
+		    const std::string& unique_tag) :
+				  Writer(dpp, y),
+				  store(_store),
+				  aio(std::move(_aio)),
+				  processor(&*aio, store,
+					    ptail_placement_rule, owner, obj_ctx,
+					    std::move(_head_obj), olh_epoch, unique_tag,
+					    dpp, y)
+  {}
```

Comment:
nit: the indentation looks off. the official [style guide](https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists) says to indent 4 spaces

Reply:
I'll fix.